languages/cpp2/src/main/java/de/jplag/cpp2/CPPTokenListener.java

```diff
@@ -0,0 +1,359 @@
+
+    @Override
+    public void exitClassSpecifier(CPP14Parser.ClassSpecifierContext context) {
+        if (context.classHead().Union() != null) {
+            addExit(CPPTokenType.UNION_END, context.getStop());
+        } else {
+            CPP14Parser.ClassKeyContext classKey = context.classHead().classKey();
+            if (classKey.Class() != null) {
+                addExit(CPPTokenType.CLASS_END, context.getStop());
+            } else if (classKey.Struct() != null) {
+                addExit(CPPTokenType.STRUCT_END, context.getStop());
+            }
+        }
+    }
+
+    @Override
+    public void enterEnumSpecifier(CPP14Parser.EnumSpecifierContext context) {
+        addEnter(CPPTokenType.ENUM_BEGIN, context.getStart());
+    }
+
+    @Override
+    public void exitEnumSpecifier(CPP14Parser.EnumSpecifierContext context) {
+        addExit(CPPTokenType.ENUM_END, context.getStop());
+    }
+
+    @Override
+    public void enterFunctionDefinition(CPP14Parser.FunctionDefinitionContext context) {
+        addEnter(CPPTokenType.FUNCTION_BEGIN, context.getStart());
+    }
+
+    @Override
+    public void exitFunctionDefinition(CPP14Parser.FunctionDefinitionContext context) {
+        addExit(CPPTokenType.FUNCTION_END, context.getStop());
+    }
+
+    @Override
+    public void enterIterationStatement(CPP14Parser.IterationStatementContext context) {
+        if (context.Do() != null) {
+            addEnter(CPPTokenType.DO_BEGIN, context.getStart());
+        } else if (context.For() != null) {
+            addEnter(CPPTokenType.FOR_BEGIN, context.getStart());
+        } else if (context.While() != null) {
+            addEnter(CPPTokenType.WHILE_BEGIN, context.getStart());
+        }
+    }
+
+    @Override
+    public void exitIterationStatement(CPP14Parser.IterationStatementContext context) {
+        if (context.Do() != null) {
+            addEnter(CPPTokenType.DO_END, context.getStop());
+        } else if (context.For() != null) {
+            addEnter(CPPTokenType.FOR_END, context.getStop());
+        } else if (context.While() != null) {
+            addEnter(CPPTokenType.WHILE_END, context.getStop());
+        }
+    }
+
+    @Override
+    public void enterSelectionStatement(CPP14Parser.SelectionStatementContext context) {
+        if (context.Switch() != null) {
+            addEnter(CPPTokenType.SWITCH_BEGIN, context.getStart());
+        } else if (context.If() != null) {
+            addEnter(CPPTokenType.IF_BEGIN, context.getStart());
+            if (context.Else() != null) {
+                addEnter(CPPTokenType.ELSE, context.Else().getSymbol());
+            }
+        }
+    }
+
+    @Override
+    public void exitSelectionStatement(CPP14Parser.SelectionStatementContext context) {
+        if (context.Switch() != null) {
+            addEnter(CPPTokenType.SWITCH_END, context.getStop());
+        } else if (context.If() != null) {
+            addEnter(CPPTokenType.IF_END, context.getStop());
+        }
+    }
```

Comment:
The grammar is the problem that leads to the weird token order.
In the grammar, an if-else combination is represented by a single selection statement, `IF statement1 ELSE statement2`.
Thus, something like this:
```cpp
if(...)
    a
else if (...)
    b
else
    c
```
is depicted as two nested selection statements:
`IF a ELSE (IF b ELSE c)`

Now, this is where things get interresting. The visitor order is not how we read this statement. The visit order is like this:

1. enter outer selection statement
2. enter inner statement
3. exit inner selection statement
3. exit outer selection statement

However, as the else is part of the selection statement, it will be either visited on entering the statement (with the if keyword and thus before the if body) or when exiting the selection statement (after the else body).

The trick to solving this is utilizing the enter/exitBlockDeclaration method and tracking the last relevant keyword. This is done in some of the newer language modules.
This solves not only the token order issue but also the issue that your if-end tokens are currently not placed correctly! They are all at the end after both selection statements. They are not closing off their if blocks.

**Rough solution (without the required imports or changes to token types), fixing token order, if-end placement and missing else_end:**

```suggestion
    @Override
    public void enterSelectionStatement(CPP14Parser.SelectionStatementContext context) {
        if (context.Switch() != null) {
            addEnter(CPPTokenType.SWITCH_BEGIN, context.getStart());
            trackedState.add(CPPTokenType.SWITCH_END);
        } else if (context.If() != null) {
            addEnter(CPPTokenType.IF_BEGIN, context.getStart());
            trackedState.add(CPPTokenType.IF_END);
            if (context.Else() != null) {
                trackedState.add(CPPTokenType.ELSE);
                trackedState.add(CPPTokenType.ELSE_END);
                lastElseToken = context.Else().getSymbol();
            }

        }
    }

    static Deque<CPPTokenType> trackedState = new ArrayDeque<>();
    static Token lastElseToken = null;

    @Override
    public void enterStatement(StatementContext context) {
        if (context.getParent().getRuleIndex() == CPP14Parser.RULE_selectionStatement) {
            if (trackedState.peekFirst() == CPPTokenType.ELSE) {
                addEnter(trackedState.removeFirst(), lastElseToken);
            }
        }
    }

    @Override
    public void exitStatement(StatementContext context) {
        if (context.getParent().getRuleIndex() == CPP14Parser.RULE_selectionStatement) {
            if (trackedState.peekFirst() == CPPTokenType.ELSE_END) {
                addExit(trackedState.removeFirst(), context.getStop());
            }
            if (trackedState.peekFirst() == CPPTokenType.IF_END) {
                addExit(trackedState.removeFirst(), context.getStop());
            }
        }
    }
```

Reply:
Thanks, I hoped to avoid additional state.

> This solves not only the token order issue but also the issue that your if-end tokens are currently not placed correctly!

I assume you're referring to the fact that multiple `}if` tokens are always placed at the very end? This is also the case in the Java language module currently.

In your code snippet, you're extracting a token for the end of an `else`, which currently isn't extracted in the Java module, but I assume it makes sense to change that too if the `}if` token is changed?

Reply:
Ah, even more issues in the Java module unearthed. Okay, then let me investigate first if the if-end and the else-end have an impact on the detection quality. If not, we do not need to fix it. I just assumed it would be correct in the Java module. I will have a look at it. Interesting!

Reply:
@SirYwell you are right, java does not place the if tokens correctly either:

```java

7               if (args == null) {
                |IF{

9               } else if (args.length > 1) {
                       |ELSE
                       |IF{

11              } else {
                       |ELSE

13              }
                |}if
                |}if
```

As a result, leave it as the Java module does it now, I can always change it in the future. However, please ensure the else tokens are correctly placed. If the else token comes before the tokens of the if block, then we cannot differ between code in the if block and code in the else block.